src/config/strgen.rs

```diff
@@ -0,0 +1,465 @@
+                self.idx += 1;
+            }
+        }
+    }
+}
+
+impl VariantExpr {
+    fn raw_iter(&self) -> VariantIter {
+        VariantIter {
+            expr: self,
+            curr: None,
+            idx: 0,
+        }
+    }
+}
+
+impl MatchExpr {
+    fn raw_iter(&self) -> SpecIter {
+        self.resolve().raw_iter()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn results_in(spec: Spec, expected: Vec<&str>) {
+        let yielded: Vec<_> = spec.into_iter().collect();
+        assert_eq!(yielded, expected);
+    }
+
+    #[test]
+    fn basic_string() {
+        results_in(
+            Spec {
+                spectype: SpecType::None,
+                string: Some("abc".to_owned()),
+            },
+            vec!["abc"],
+        );
+    }
+
+    #[test]
+    fn basic_variant() {
+        results_in(
+            Spec {
+                string: Some("a".to_owned()),
+                spectype: SpecType::Variant(
+                    Box::new(VariantExpr {
+                        specs: vec![
+                            Spec {
+                                spectype: SpecType::None,
+                                string: Some("b".to_owned()),
+                            },
+                            Spec {
+                                spectype: SpecType::None,
+                                string: Some("c".to_owned()),
+                            },
+                        ],
+                    }),
+                    None,
+                ),
+            },
+            vec!["ab", "ac"],
+        )
+    }
+
+    #[test]
+    fn basic_match() {
+        results_in(
+            // Equivalent to `d{ incorrect-os: g, default: e }f`.
+            Spec {
+                string: Some("d".to_owned()),
+                spectype: SpecType::Match(
+                    Box::new(MatchExpr {
+                        cases: vec![(
+                            Expr {
+                                exprtype: if cfg!(windows) {
+                                    ExprType::Linux
+                                } else {
+                                    ExprType::Windows
+                                },
+                            },
+                            Spec {
+                                string: Some("g".to_owned()),
+                                spectype: SpecType::None,
+                            },
+                        )],
+                        default: Spec {
+                            string: Some("e".to_owned()),
+                            spectype: SpecType::None,
+                        },
+                    }),
+                    Some(Box::new(Spec {
+                        string: Some("f".to_owned()),
+                        spectype: SpecType::None,
+                    })),
+                ),
+            },
+            vec!["def"],
+        )
+    }
+
+    #[test]
+    fn nested_variant() {
+        results_in(
```

Comment:
The input format of the unit tests is over-complicated. A more human-readable input format should be implemented.